ui/src/app/components/search-input.tsx

```diff
@@ -0,0 +1,225 @@
+import {TextInput} from 'app/components/inputs';
+import {TooltipTrigger} from 'app/components/popups';
+import colors, {colorWithWhiteness} from 'app/styles/colors';
+import {reactStyles} from 'app/utils';
+import * as fp from 'lodash/fp';
+import * as React from 'react';
+
+const styles = reactStyles({
+  dropdownMenu: {
+    display: 'block',
+    maxHeight: '12rem',
+    minHeight: '30px',
+    visibility: 'visible',
+    overflowY: 'scroll',
+    width: '35%',
+    marginTop: '.25rem',
+    zIndex: 100,
+    border: '1px solid #979797',
+    borderRadius: '5px',
+    backgroundColor: colors.white,
+  },
+  open: {
+    position: 'absolute',
+    backgroundColor: colors.white,
+    border: '1px solid'
+  },
+  box: {
+    borderRadius: '5px',
+    paddingTop: '0.2rem',
+    paddingLeft: '0.2rem',
+    color: colors.primary,
+  },
+  boxHover: {
+    background: 'rgb(234, 243, 250)',
+    color: '#262262',
+    paddingTop: '0.2rem',
+    paddingLeft: '0.2rem',
+  },
+  boxHoverElement: {
+    color: colors.primary,
+    margin: 0,
+    padding: 0,
+    fontFamily: 'Montserrat',
+    fontSize: '14px',
+    lineHeight: '32px',
+  },
+});
+
+// Fifo accepts a function 'fn' that returns a promise, and returns a
+// new function that returns a promise. Promises returned by the
+// resulting function return their result of the last-instantiated
+// promise. For example, if a0, a1 are the start and end times of a
+// computation represented by promise A, and b0, b1 are the start and end
+// times of a computation represented by promise B, and a0 < b0 < b1 <
+// a1, then the value returned by promise A at time a1 will be the result
+// of promise B, *not* promise A, because B was instantiated after A.
+function fifo(fn) {
+  let nextCallID = 1;
+  let lastCompletedCallID = 0;
+  let lastCompletedResult = null;
+  return (...args) => {
+    const thisCallID = nextCallID++;
+    return new Promise((accept, reject) => {
+      fn.apply(null, args).then((newResult) => {
+        if (thisCallID > lastCompletedCallID) {
+          lastCompletedCallID = thisCallID;
+          lastCompletedResult = newResult;
+        }
+        accept(lastCompletedResult);
+      }).catch((e) => {
+        reject(e);
+      });
+    });
+  };
+}
+
+export interface SearchInputState {
+  matches: Array<string>;
+  hover: Array<boolean>;
+  state: number;
+}
+
+export interface SearchInputProps {
+  enabled: boolean;
+  placeholder: string;
+  value: string;
+  onSearch: Function; // Returns a promise
```

Comment:
I believe you can type function parameters and returns, so you can turn that comment into code!

Reply:
Fixed